optaplanner-core/src/main/java/org/optaplanner/core/api/domain/variable/IndexShadowVariable.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.api.domain.variable;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.optaplanner.core.api.domain.entity.PlanningEntity;
+import org.optaplanner.core.api.solver.Solver;
+
+/**
+ * Specifies that a bean property (or a field) is an index of this planning value in another entity's
+ * {@link PlanningListVariable}.
+ * <p>
+ * It is specified on a getter of a java bean property (or a field) of a {@link PlanningEntity} class.
+ * <p>
+ * The source variable must be a {@link PlanningListVariable list variable}.
+ */
+@Target({ METHOD, FIELD })
+@Retention(RUNTIME)
+public @interface IndexShadowVariable {
+
+    /**
+     * The source variable must be a {@link PlanningListVariable list variable}.
+     * <p>
+     * When the {@link Solver} changes a genuine variable, it adjusts the shadow variable accordingly.
+     * In practice, the {@link Solver} ignores shadow variables (except for consistency housekeeping).
+     *
+     * @return property name of the list variable that contains instances of this planning value
+     */
+    String sourceVariableName();
```

Comment:
What happens if I have two planning entities, A and B, with a planning list variable of the same name, which both take values from the same class. (ex: Bus and Taxi both have a Location List).

Reply:
It will fail fast if among all entity classes there are two or more planning variables using the same variable name. So the following is impossible because `locationList` is not a unique variable name (pseudo code):
```java
@PlanningListVariable Taxi.locationList: List<Location>
@PlanningListVariable Bus.locationList: List<Location>

@IndeShadowVariable(source="locationList") Location.index: Integer
```
I think this can never work.

(TODO this should probably fail fast even without using index shadow variable. A given planning value instance must only be used in a single list variable. In other words, one value range can only be used by one list variable and multiple value ranges must not overlap.)

The following approach _should_ work now:
```java
@PlanningListVariable Taxi.taxiLocationList: List<TaxiLocation>
@PlanningListVariable Bus.busLocationList: List<BusLocation>

@IndeShadowVariable(source="taxiLocationList") TaxiLocation.index: Integer
@IndeShadowVariable(source="busLocationList") BusLocation.index: Integer

TaxiLocation extends Location, BusLocation extends Location
```
I _could_ also relax the requirement a bit and only fail if two variables have the same name and the same element type. Then, also this would be possible:
```java
@PlanningListVariable Taxi.locationList: List<TaxiLocation>
@PlanningListVariable Bus.locationList: List<BusLocation>

@IndeShadowVariable(source="taxiLocationList") TaxiLocation.index: Integer
@IndeShadowVariable(source="busLocationList") BusLocation.index: Integer

TaxiLocation extends Location, BusLocation extends Location
```

I believe we can afford to put this on the backlog and defer it until there's a real demand for it.